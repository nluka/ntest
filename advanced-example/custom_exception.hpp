#ifndef CUSTOM_EXCEPTION_HPP
#define CUSTOM_EXCEPTION_HPP

#include <exception>

struct custom_exception : public std::exception
{
  [[nodiscard]] virtual char const *what() const noexcept
  {
    return "Something bad happened";
  }
};

#endif // CUSTOM_EXCEPTION_HPP
