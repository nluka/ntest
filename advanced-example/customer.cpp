#include "customer.hpp"

bool customer::operator!=(customer const &other) const
{
  return this->first_name != other.first_name
    || this->last_name != other.last_name
    || this->age != other.age;
}

std::ostream& operator<<(std::ostream &os, customer const &cust)
{
  os
    << '(' << cust.first_name << ' ' << cust.last_name
    << ", " << (int)cust.age << ')';
  return os;
}
