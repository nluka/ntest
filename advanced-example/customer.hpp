#ifndef EXAMPLE_CUSTOMER_HPP
#define EXAMPLE_CUSTOMER_HPP

#include <string>
#include <ostream>

struct customer
{
  std::string first_name;
  std::string last_name;
  uint8_t age;

  bool operator!=(customer const &) const;
  friend std::ostream& operator<<(std::ostream &, customer const &);
};

#endif // EXAMPLE_CUSTOMER_HPP
