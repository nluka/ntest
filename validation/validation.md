# validation

Wed Mar 22 19:31:00 2023

60 failed

41 passed

| | Type | Expected | Actual | Location (fn:ln,col) | Source File |
| - | - | - | - | - | - |
| ❌ | bool | true | false | int main():58,23 | validation.cpp |
| ❌ | signed char | -128 | -28 | int main():62,23 | validation.cpp |
| ❌ | signed char | 0 | 8 | int main():64,23 | validation.cpp |
| ❌ | signed char | 127 | 105 | int main():66,23 | validation.cpp |
| ❌ | unsigned char | 0 | 1 | int main():70,24 | validation.cpp |
| ❌ | unsigned char | 255 | 224 | int main():72,24 | validation.cpp |
| ❌ | short | -32768 | -32668 | int main():76,24 | validation.cpp |
| ❌ | short | 0 | 42 | int main():78,24 | validation.cpp |
| ❌ | short | 32767 | 32756 | int main():80,24 | validation.cpp |
| ❌ | unsigned short | 0 | 1 | int main():84,25 | validation.cpp |
| ❌ | unsigned short | 65535 | 65091 | int main():86,25 | validation.cpp |
| ❌ | int | -2147483648 | -2147483548 | int main():90,24 | validation.cpp |
| ❌ | int | 0 | 42 | int main():92,24 | validation.cpp |
| ❌ | int | 2147483647 | 2147483646 | int main():94,24 | validation.cpp |
| ❌ | unsigned int | 0 | 1 | int main():98,25 | validation.cpp |
| ❌ | unsigned int | 4294967295 | 4294966851 | int main():100,25 | validation.cpp |
| ❌ | long | -9223372036854775808 | -9223372036854775708 | int main():104,24 | validation.cpp |
| ❌ | long | 0 | 42 | int main():106,24 | validation.cpp |
| ❌ | long | 9223372036854775807 | 9223372036854774806 | int main():108,24 | validation.cpp |
| ❌ | unsigned long | 0 | 1 | int main():112,25 | validation.cpp |
| ❌ | unsigned long | 18446744073709551615 | 18446744073709551573 | int main():114,25 | validation.cpp |
| ❌ | char* | [validation.cpp@121,25.expected](validation.cpp@121,25.expected) | [validation.cpp@121,25.actual](validation.cpp@121,25.actual) | int main():121,25 | validation.cpp |
| ❌ | char* | [validation.cpp@122,25.expected](validation.cpp@122,25.expected) | [validation.cpp@122,25.actual](validation.cpp@122,25.actual) | int main():122,25 | validation.cpp |
| ❌ | char* | [validation.cpp@124,25.expected](validation.cpp@124,25.expected) | [validation.cpp@124,25.actual](validation.cpp@124,25.actual) | int main():124,25 | validation.cpp |
| ❌ | char* | [validation.cpp@132,27.expected](validation.cpp@132,27.expected) | [validation.cpp@132,27.actual](validation.cpp@132,27.actual) | int main():132,27 | validation.cpp |
| ❌ | char* | [validation.cpp@133,27.expected](validation.cpp@133,27.expected) | [validation.cpp@133,27.actual](validation.cpp@133,27.actual) | int main():133,27 | validation.cpp |
| ❌ | char* | [validation.cpp@142,27.expected](validation.cpp@142,27.expected) | [validation.cpp@142,27.actual](validation.cpp@142,27.actual) | int main():142,27 | validation.cpp |
| ❌ | char* | [validation.cpp@151,25.expected](validation.cpp@151,25.expected) | [validation.cpp@151,25.actual](validation.cpp@151,25.actual) | int main():151,25 | validation.cpp |
| ❌ | char* | [validation.cpp@152,27.expected](validation.cpp@152,27.expected) | [validation.cpp@152,27.actual](validation.cpp@152,27.actual) | int main():152,27 | validation.cpp |
| ❌ | int [] | [validation.cpp@161,24.expected](validation.cpp@161,24.expected) | [validation.cpp@161,24.actual](validation.cpp@161,24.actual) | int main():161,24 | validation.cpp |
| ❌ | int [] | [validation.cpp@162,24.expected](validation.cpp@162,24.expected) | [validation.cpp@162,24.actual](validation.cpp@162,24.actual) | int main():162,24 | validation.cpp |
| ❌ | customer [] | [validation.cpp@189,24.expected](validation.cpp@189,24.expected) | [validation.cpp@189,24.actual](validation.cpp@189,24.actual) | int main():189,24 | validation.cpp |
| ❌ | customer [] | [validation.cpp@190,24.expected](validation.cpp@190,24.expected) | [validation.cpp@190,24.actual](validation.cpp@190,24.actual) | int main():190,24 | validation.cpp |
| ❌ | customer [] | [validation.cpp@191,24.expected](validation.cpp@191,24.expected) | [validation.cpp@191,24.actual](validation.cpp@191,24.actual) | int main():191,24 | validation.cpp |
| ❌ | std::vector\<int\> | [validation.cpp@201,27.expected](validation.cpp@201,27.expected) | [validation.cpp@201,27.actual](validation.cpp@201,27.actual) | int main():201,27 | validation.cpp |
| ❌ | std::vector\<int\> | [validation.cpp@202,27.expected](validation.cpp@202,27.expected) | [validation.cpp@202,27.actual](validation.cpp@202,27.actual) | int main():202,27 | validation.cpp |
| ❌ | std::vector\<customer\> | [validation.cpp@229,27.expected](validation.cpp@229,27.expected) | [validation.cpp@229,27.actual](validation.cpp@229,27.actual) | int main():229,27 | validation.cpp |
| ❌ | std::vector\<customer\> | [validation.cpp@230,27.expected](validation.cpp@230,27.expected) | [validation.cpp@230,27.actual](validation.cpp@230,27.actual) | int main():230,27 | validation.cpp |
| ❌ | std::vector\<customer\> | [validation.cpp@231,27.expected](validation.cpp@231,27.expected) | [validation.cpp@231,27.actual](validation.cpp@231,27.actual) | int main():231,27 | validation.cpp |
| ❌ | std::vector\<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>\> | [validation.cpp@256,27.expected](validation.cpp@256,27.expected) | [validation.cpp@256,27.actual](validation.cpp@256,27.actual) | int main():256,27 | validation.cpp |
| ❌ | std::vector\<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>\> | [validation.cpp@257,27.expected](validation.cpp@257,27.expected) | [validation.cpp@257,27.actual](validation.cpp@257,27.actual) | int main():257,27 | validation.cpp |
| ❌ | std::array\<int, 4\> | [validation.cpp@267,27.expected](validation.cpp@267,27.expected) | [validation.cpp@267,27.actual](validation.cpp@267,27.actual) | int main():267,27 | validation.cpp |
| ❌ | std::array\<int, 4\> | [validation.cpp@268,27.expected](validation.cpp@268,27.expected) | [validation.cpp@268,27.actual](validation.cpp@268,27.actual) | int main():268,27 | validation.cpp |
| ❌ | std::array\<customer, 3\> | [validation.cpp@290,27.expected](validation.cpp@290,27.expected) | [validation.cpp@290,27.actual](validation.cpp@290,27.actual) | int main():290,27 | validation.cpp |
| ❌ | std::array\<customer, 3\> | [validation.cpp@291,27.expected](validation.cpp@291,27.expected) | [validation.cpp@291,27.actual](validation.cpp@291,27.actual) | int main():291,27 | validation.cpp |
| ❌ | std::array\<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>, 4\> | [validation.cpp@315,27.expected](validation.cpp@315,27.expected) | [validation.cpp@315,27.actual](validation.cpp@315,27.actual) | int main():315,27 | validation.cpp |
| ❌ | std::array\<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>, 4\> | [validation.cpp@316,27.expected](validation.cpp@316,27.expected) | [validation.cpp@316,27.actual](validation.cpp@316,27.actual) | int main():316,27 | validation.cpp |
| ❌ | text file | [./files/expected.txt](./files/expected.txt) | [./files/a.txt](./files/a.txt) | int main():321,28 | validation.cpp |
| ❌ | text file | [./files/expected.txt](./files/expected.txt) | [./files/b.txt](./files/b.txt) | int main():322,28 | validation.cpp |
| ❌ | text file | file not found | [./files/a.txt](./files/a.txt) | int main():324,28 | validation.cpp |
| ❌ | text file | [./files/expected.txt](./files/expected.txt) | file not found | int main():325,28 | validation.cpp |
| ❌ | binary file | [./files/expected.bin](./files/expected.bin) | [./files/a.bin](./files/a.bin) | int main():328,30 | validation.cpp |
| ❌ | binary file | [./files/expected.bin](./files/expected.bin) | [./files/b.bin](./files/b.bin) | int main():329,30 | validation.cpp |
| ❌ | binary file | file not found | [./files/a.bin](./files/a.bin) | int main():331,30 | validation.cpp |
| ❌ | binary file | [./files/expected.bin](./files/expected.bin) | file not found | int main():332,30 | validation.cpp |
| ❌ | char* | [validation.cpp@344,25.expected](validation.cpp@344,25.expected) | [validation.cpp@344,25.actual](validation.cpp@344,25.actual) | int main():344,25 | validation.cpp |
| ❌ | std::runtime_error | to be thrown | different exception type was thrown | int main():346,58 | validation.cpp |
| ❌ | std::runtime_error | to be thrown | nothing was thrown | int main():352,58 | validation.cpp |
| ❌ | custom_exception | to be thrown | nothing was thrown | int main():365,56 | validation.cpp |
| ❌ | custom_exception | to be thrown | different exception type was thrown | int main():372,56 | validation.cpp |

| | Type | Expected | Location (fn:ln,col) | Source File |
| - | - | - | - | - |
| ✅ | bool | true | int main():57,23 | validation.cpp |
| ✅ | signed char | -128 | int main():61,23 | validation.cpp |
| ✅ | signed char | 0 | int main():63,23 | validation.cpp |
| ✅ | signed char | 127 | int main():65,23 | validation.cpp |
| ✅ | unsigned char | 0 | int main():69,24 | validation.cpp |
| ✅ | unsigned char | 255 | int main():71,24 | validation.cpp |
| ✅ | short | -32768 | int main():75,24 | validation.cpp |
| ✅ | short | 0 | int main():77,24 | validation.cpp |
| ✅ | short | 32767 | int main():79,24 | validation.cpp |
| ✅ | unsigned short | 0 | int main():83,25 | validation.cpp |
| ✅ | unsigned short | 65535 | int main():85,25 | validation.cpp |
| ✅ | int | -2147483648 | int main():89,24 | validation.cpp |
| ✅ | int | 0 | int main():91,24 | validation.cpp |
| ✅ | int | 2147483647 | int main():93,24 | validation.cpp |
| ✅ | unsigned int | 0 | int main():97,25 | validation.cpp |
| ✅ | unsigned int | 4294967295 | int main():99,25 | validation.cpp |
| ✅ | long | -9223372036854775808 | int main():103,24 | validation.cpp |
| ✅ | long | 0 | int main():105,24 | validation.cpp |
| ✅ | long | 9223372036854775807 | int main():107,24 | validation.cpp |
| ✅ | unsigned long | 0 | int main():111,25 | validation.cpp |
| ✅ | unsigned long | 18446744073709551615 | int main():113,25 | validation.cpp |
| ✅ | char* | len=12 <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;'>Hello\nworl</span> *... 2 more* | int main():123,25 | validation.cpp |
| ✅ | char* | len=26 <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;'>abcdefghij</span> *... 16 more* | int main():125,25 | validation.cpp |
| ✅ | char* | len=26 <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;'>Testing co</span> *... 16 more* | int main():134,27 | validation.cpp |
| ✅ | char* | len=26 <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;'>Testing\0co</span> *... 16 more* | int main():143,27 | validation.cpp |
| ✅ | char* | len=10 <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;'>\`Backtick\`</span> | int main():149,25 | validation.cpp |
| ✅ | char* | len=10 <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;'>\`Backtick\`</span> | int main():150,27 | validation.cpp |
| ✅ | int [] | sz=4 __[__ <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 0'>0</span>, <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 1'>1</span>,  *... 2 more* __]__ | int main():163,24 | validation.cpp |
| ✅ | customer [] | sz=3 __[__ <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 0'>(Alice Smith, 22)</span>, <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 1'>(Joe Schmoe, 100)</span>,  *... 1 more* __]__ | int main():192,24 | validation.cpp |
| ✅ | std::vector\<int\> | sz=4 __[__ <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 0'>0</span>, <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 1'>1</span>,  *... 2 more* __]__ | int main():203,27 | validation.cpp |
| ✅ | std::vector\<customer\> | sz=3 __[__ <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 0'>(Alice Smith, 22)</span>, <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 1'>(Joe Schmoe, 100)</span>,  *... 1 more* __]__ | int main():232,27 | validation.cpp |
| ✅ | std::vector\<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>\> | sz=4 __[__ <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 0'>Testing</span>, <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 1'>code</span>,  *... 2 more* __]__ | int main():258,27 | validation.cpp |
| ✅ | std::array\<int, 4\> | sz=4 __[__ <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 0'>0</span>, <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 1'>1</span>,  *... 2 more* __]__ | int main():269,27 | validation.cpp |
| ✅ | std::array\<customer, 3\> | sz=3 __[__ <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 0'>(Alice Smith, 22)</span>, <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 1'>(Joe Schmoe, 100)</span>,  *... 1 more* __]__ | int main():292,27 | validation.cpp |
| ✅ | std::array\<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>, 4\> | sz=4 __[__ <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 0'>Testing</span>, <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;' title='index 1'>code</span>,  *... 2 more* __]__ | int main():317,27 | validation.cpp |
| ✅ | text file | [./files/expected.txt](./files/expected.txt) | int main():323,28 | validation.cpp |
| ✅ | binary file | [./files/expected.bin](./files/expected.bin) | int main():330,30 | validation.cpp |
| ✅ | std::runtime_error | to be thrown | int main():338,58 | validation.cpp |
| ✅ | char* | len=23 <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;'>some expla</span> *... 13 more* | int main():343,25 | validation.cpp |
| ✅ | custom_exception | to be thrown | int main():359,56 | validation.cpp |
| ✅ | char* | len=22 <span style='border: 1px solid darkgray;font-family: monospace;padding: 2px;white-space: pre-wrap;'>Something </span> *... 12 more* | int main():363,25 | validation.cpp |

