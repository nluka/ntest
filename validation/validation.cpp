#include <cassert>
#include <iostream>

#include "customer.hpp"
#include "custom_exception.hpp"
#include "timer.hpp"
#include "../ntest.hpp"

template <typename Ty, size_t Length>
consteval
size_t len(Ty (&)[Length])
{
  return Length;
}

int main()
{
  ntest::config::set_show_column_numbers(false);

  size_t observed_pass_cnt = 0, observed_fail_cnt = 0;

  auto const ensure_passed = [&observed_pass_cnt](
    // not actually used, but for code annotation:
    [[maybe_unused]] char const *const desc = nullptr,
    std::source_location const loc = std::source_location::current()
  ) {
    ++observed_pass_cnt;
    size_t const ntest_pass_cnt = ntest::pass_count();
    if (ntest_pass_cnt != observed_pass_cnt) {
      std::cerr << "ERROR: incorrect assertion outcome (wanted pass, got fail) on line " << loc.line() << '\n';
      ntest::generate_report("validation");
      std::exit(1);
    }
  };

  auto const ensure_failed = [&observed_fail_cnt](
    // not actually used, but for code annotation:
    [[maybe_unused]] char const *const desc = nullptr,
    std::source_location const loc = std::source_location::current()
  ) {
    ++observed_fail_cnt;
    size_t const ntest_fail_cnt = ntest::fail_count();
    if (ntest_fail_cnt != observed_fail_cnt) {
      std::cerr << "ERROR: incorrect assertion outcome (wanted fail, got pass) on line " << loc.line() << '\n';
      ntest::generate_report("validation");
      std::exit(1);
    }
  };

  try
  {
    timer::scoped<timer::unit::MILLISECONDS> timer("execution", std::cout);

    ntest::init();
    ntest::config::set_max_arr_preview_len(2);
    ntest::config::set_max_str_preview_len(15);

    // bool
    ntest::assert_bool(true, true); ensure_passed();
    ntest::assert_bool(true, false); ensure_failed();

    // int8
    ntest::assert_int8(INT8_MIN, INT8_MIN); ensure_passed();
    ntest::assert_int8(INT8_MIN, INT8_MIN + 100); ensure_failed();
    ntest::assert_int8(0, 0); ensure_passed();
    ntest::assert_int8(0, 8); ensure_failed();
    ntest::assert_int8(INT8_MAX, INT8_MAX); ensure_passed();
    ntest::assert_int8(INT8_MAX, INT8_MAX - 22); ensure_failed();

    // uint8
    ntest::assert_uint8(0, 0); ensure_passed();
    ntest::assert_uint8(0, 1); ensure_failed();
    ntest::assert_uint8(UINT8_MAX, UINT8_MAX); ensure_passed();
    ntest::assert_uint8(UINT8_MAX, UINT8_MAX - 31); ensure_failed();

    // int16
    ntest::assert_int16(INT16_MIN, INT16_MIN); ensure_passed();
    ntest::assert_int16(INT16_MIN, INT16_MIN + 100); ensure_failed();
    ntest::assert_int16(0, 0); ensure_passed();
    ntest::assert_int16(0, 42); ensure_failed();
    ntest::assert_int16(INT16_MAX, INT16_MAX); ensure_passed();
    ntest::assert_int16(INT16_MAX, INT16_MAX - 11); ensure_failed();

    // uint16
    ntest::assert_uint16(0, 0); ensure_passed();
    ntest::assert_uint16(0, 1); ensure_failed();
    ntest::assert_uint16(UINT16_MAX, UINT16_MAX); ensure_passed();
    ntest::assert_uint16(UINT16_MAX, UINT16_MAX - 444); ensure_failed();

    // int32
    ntest::assert_int32(INT32_MIN, INT32_MIN); ensure_passed();
    ntest::assert_int32(INT32_MIN, INT32_MIN + 100); ensure_failed();
    ntest::assert_int32(0, 0); ensure_passed();
    ntest::assert_int32(0, 42); ensure_failed();
    ntest::assert_int32(INT32_MAX, INT32_MAX); ensure_passed();
    ntest::assert_int32(INT32_MAX, INT32_MAX - 1); ensure_failed();

    // uint32
    ntest::assert_uint32(0, 0); ensure_passed();
    ntest::assert_uint32(0, 1); ensure_failed();
    ntest::assert_uint32(UINT32_MAX, UINT32_MAX); ensure_passed();
    ntest::assert_uint32(UINT32_MAX, UINT32_MAX - 444); ensure_failed();

    // int64
    ntest::assert_int64(INT64_MIN, INT64_MIN); ensure_passed();
    ntest::assert_int64(INT64_MIN, INT64_MIN + 100); ensure_failed();
    ntest::assert_int64(0, 0); ensure_passed();
    ntest::assert_int64(0, 42); ensure_failed();
    ntest::assert_int64(INT64_MAX, INT64_MAX); ensure_passed();
    ntest::assert_int64(INT64_MAX, INT64_MAX - 1001); ensure_failed();

    // uint64_t
    ntest::assert_uint64(0, 0); ensure_passed();
    ntest::assert_uint64(0, 1); ensure_failed();
    ntest::assert_uint64(UINT64_MAX, UINT64_MAX); ensure_passed();
    ntest::assert_uint64(UINT64_MAX, UINT64_MAX - 42); ensure_failed();

    // char*
    {
      char const *const hello_world = "Hello\nworld!";
      char const *const alphabet = "abcdefghijklmnopqrstuvwxyz";

      ntest::assert_cstr(hello_world, "hello\nworld!"); ensure_failed();
      ntest::assert_cstr(hello_world, "Hello world"); ensure_failed();
      ntest::assert_cstr(hello_world, hello_world); ensure_passed();
      ntest::assert_cstr(hello_world, alphabet); ensure_failed();
      ntest::assert_cstr(alphabet, alphabet); ensure_passed();
    }

    // std::string
    {
      std::string const expected = "Testing code is important.";

      ntest::assert_stdstr(expected, "Testing code is not important."); ensure_failed();
      ntest::assert_stdstr(expected, "What is testing?"); ensure_failed();
      ntest::assert_stdstr(expected, expected); ensure_passed();
    }

    // std::string with NULs
    {
      std::stringstream ss{};
      ss << "Testing" << '\0' << "code" << '\0' << "is" << '\0' << "important.";

      ntest::assert_stdstr("Testing code is important.", ss.str()); ensure_failed();
      ntest::assert_stdstr(ss.str(), ss.str()); ensure_passed();
    }

    // strings with backticks
    {
      char const *const str = "`Backtick`";
      ntest::assert_cstr(str, str); ensure_passed();
      ntest::assert_stdstr(str, str); ensure_passed();
      ntest::assert_cstr(str, "nonsense"); ensure_failed();
      ntest::assert_stdstr(str, "nonsense"); ensure_failed();
    }
    {
      std::vector<std::string> const a { "`Backtick`" };
      ntest::assert_stdvec(a, a); ensure_passed();
      ntest::assert_stdvec(a, {}); ensure_failed();
    }

    // int []
    {
      int const expected[] { 0, 1, 2, 3 };
      int const a[]        { 1, 2, 3, 4 };
      int const b[]        { 0, 0, 0, 0 };

      ntest::assert_arr(expected, len(expected), a, len(a)); ensure_failed();
      ntest::assert_arr(expected, len(expected), b, len(b)); ensure_failed();
      ntest::assert_arr(expected, len(expected), expected, len(expected)); ensure_passed();
    }

    // customer []
    {
      customer const expected[] {
        {"Alice", "Smith", 22},
        {"Joe", "Schmoe", 100},
        {"Rob", "Robertson", 53},
      };
      customer const a[] {
        {"Alice", "Willians", 22},
        {"Will", "Robertson", 54},
      };
      customer const b[] {
        {"Joe", "Schmoe", 100},
        {"Rob", "Robertson", 53},
        {"Alice", "Smith", 22},
      };
      customer const c[] {
        {"Alice", "Smith", 22},
        {"Joe", "Schmoe", 100},
        {"Rob", "Robertson", 53},
        {"John", "Miller", 67},
      };

      ntest::assert_arr(expected, len(expected), a, len(a)); ensure_failed();
      ntest::assert_arr(expected, len(expected), b, len(b)); ensure_failed();
      ntest::assert_arr(expected, len(expected), c, len(c)); ensure_failed();
      ntest::assert_arr(expected, len(expected), expected, len(expected)); ensure_passed();
    }

    // std::vector<int>
    {
      std::vector<int> const expected { 0, 1, 2, 3 };
      std::vector<int> const a        { 1, 2, 3, 4 };
      std::vector<int> const b        { 0, 0, 0, 0 };

      ntest::assert_stdvec(expected, a); ensure_failed();
      ntest::assert_stdvec(expected, b); ensure_failed();
      ntest::assert_stdvec(expected, expected); ensure_passed();
    }

    // std::vector<customer>
    {
      std::vector<customer> const expected {
        {"Alice", "Smith", 22},
        {"Joe", "Schmoe", 100},
        {"Rob", "Robertson", 53},
      };
      std::vector<customer> const a {
        {"Alice", "Willians", 22},
        {"Will", "Robertson", 54},
      };
      std::vector<customer> const b {
        {"Joe", "Schmoe", 100},
        {"Rob", "Robertson", 53},
        {"Alice", "Smith", 22},
      };
      std::vector<customer> const c {
        {"Alice", "Smith", 22},
        {"Joe", "Schmoe", 100},
        {"Rob", "Robertson", 53},
        {"John", "Miller", 67},
      };

      ntest::assert_stdvec(expected, a); ensure_failed();
      ntest::assert_stdvec(expected, b); ensure_failed();
      ntest::assert_stdvec(expected, c); ensure_failed();
      ntest::assert_stdvec(expected, expected); ensure_passed();
    }

    // std::vector<std::string>
    {
      std::vector<std::string> const expected {
        "Testing",
        "code",
        "is",
        "important!",
      };
      std::vector<std::string> const a {
        "Testing",
        "code",
        "is",
        "not",
        "important.",
      };
      std::vector<std::string> const b {
        "What",
        "is",
        "ntesting?",
      };

      ntest::assert_stdvec(expected, a); ensure_failed();
      ntest::assert_stdvec(expected, b); ensure_failed();
      ntest::assert_stdvec(expected, expected); ensure_passed();

      std::vector<std::string> const c {
        "Testing\n",
        "code\n",
        "is\n",
        "important!\n",
      };
      std::vector<std::string> const d {
        "`Testing`",
        "`code`",
        "`is`",
        "`important!`",
      };

      ntest::assert_stdvec(c, c); ensure_passed(); // check preview has \n escaped
      ntest::assert_stdvec(expected, c); ensure_failed(); // check .actual has \n escaped
      ntest::assert_stdvec(expected, d); ensure_failed(); // check .actual has does not escape `
    }

    // std::array<int>
    {
      std::array<int, 4> const expected { 0, 1, 2, 3 };
      std::array<int, 4> const a        { 1, 2, 3, 4 };
      std::array<int, 4> const b        { 0, 0, 0, 0 };

      ntest::assert_stdarr(expected, a); ensure_failed();
      ntest::assert_stdarr(expected, b); ensure_failed();
      ntest::assert_stdarr(expected, expected); ensure_passed();
    }

    // std::array<customer>
    {
      std::array<customer, 3> const expected {
        customer("Alice", "Smith", 22),
        customer("Joe", "Schmoe", 100),
        customer("Rob", "Robertson", 53),
      };
      std::array<customer, 3> const a {
        customer("Smith", "Alice", 22),
        customer("Rob", "Robertson", 53),
        customer("Joe", "Schmoe", 100),
      };
      std::array<customer, 3> const b {
        customer("Alice", "Smith", 22),
        customer("Joe", "Schmoe", 77),
        customer("Rob", "Robertson", 40),
      };

      ntest::assert_stdarr(expected, a); ensure_failed();
      ntest::assert_stdarr(expected, b); ensure_failed();
      ntest::assert_stdarr(expected, expected); ensure_passed();
    }

    // std::array<std::string>
    {
      std::array<std::string, 4> const expected {
        "Testing",
        "code",
        "is",
        "important!",
      };
      std::array<std::string, 4> const a {
        "Testing",
        "code",
        "aint",
        "important.",
      };
      std::array<std::string, 4> const b {
        "What",
        "is",
        "ntesting?",
      };

      ntest::assert_stdarr(expected, a); ensure_failed();
      ntest::assert_stdarr(expected, b); ensure_failed();
      ntest::assert_stdarr(expected, expected); ensure_passed();
    }

    // text files
    ntest::assert_text_file("./files/expected.txt", "./files/a.txt"); ensure_failed();
    ntest::assert_text_file("./files/expected.txt", "./files/b.txt"); ensure_failed();
    ntest::assert_text_file("./files/expected.txt", "./files/expected.txt"); ensure_passed();
    ntest::assert_text_file("./files/does_not_exist.txt", "./files/a.txt"); ensure_failed();
    ntest::assert_text_file("./files/expected.txt", "./files/does_not_exist.txt"); ensure_failed();

    // binary files
    ntest::assert_binary_file("./files/expected.bin", "./files/a.bin"); ensure_failed();
    ntest::assert_binary_file("./files/expected.bin", "./files/b.bin"); ensure_failed();
    ntest::assert_binary_file("./files/expected.bin", "./files/expected.bin"); ensure_passed();
    ntest::assert_binary_file("./files/does_not_exist.bin", "./files/a.bin"); ensure_failed();
    ntest::assert_binary_file("./files/expected.bin", "./files/does_not_exist.bin"); ensure_failed();

    // exceptions
    {
      std::string what_str;

      what_str = ntest::assert_throws<std::runtime_error>([]() {
        throw std::runtime_error(std::string("some explanatory string"));
      });
      assert(!what_str.empty());
      ensure_passed("because correct exception type was thrown");
      ntest::assert_cstr("some explanatory string", what_str.c_str()); ensure_passed();
      ntest::assert_cstr("some other explanatory string", what_str.c_str()); ensure_failed();

      what_str = ntest::assert_throws<std::runtime_error>([]() {
        throw std::exception();
      });
      assert(what_str.empty());
      ensure_failed("because wrong exception type was thrown");

      what_str = ntest::assert_throws<std::runtime_error>([]() {
        int a = 10, b = 20;
        [[maybe_unused]] int c = a + b;
      });
      assert(what_str.empty());
      ensure_failed("because nothing was thrown");

      what_str = ntest::assert_throws<custom_exception>([]() {
        throw custom_exception();
      });
      ensure_passed("because correct exception type was thrown");
      ntest::assert_cstr("Something bad happened", what_str.c_str()); ensure_passed();

      what_str = ntest::assert_throws<custom_exception>([]() {
        int a = 10, b = 20;
        [[maybe_unused]] int c = a + b;
      });
      assert(what_str.empty());
      ensure_failed("because nothing was thrown");

      what_str = ntest::assert_throws<custom_exception>([]() {
        throw std::exception();
      });
      assert(what_str.empty());
      ensure_failed("because wrong exception type was thrown");
    }

    try
    {
      int expected[] = { 1, 2, 3 };
      int *actual = nullptr;
      // this should throw because we said `actual` has 3 elems
      // when it's really a nullptr:
      ntest::assert_arr(expected, 3, actual, 3);
      std::cerr << "failed to catch user error!\n";
    }
    catch (std::runtime_error const &err)
    {
      std::cout << "successfully caught user error: " << err.what() << '\n';
    }

    // verify pass/fail counts:
    {
      size_t const
        expected_pass_cnt = observed_pass_cnt,
        expected_fail_cnt = observed_fail_cnt,
        actual_pass_cnt = ntest::pass_count(),
        actual_fail_cnt = ntest::fail_count()
      ;
      bool const success =
        (actual_pass_cnt == expected_pass_cnt) &&
        (actual_fail_cnt == expected_fail_cnt)
      ;
      std::cout
        << (success ? "SUCCESS" : "FAIL") << "! "
        << "expected pass/fail = " << expected_pass_cnt << '/' << expected_fail_cnt
        << ", observed " << actual_pass_cnt << '/' << actual_fail_cnt << '\n';
      ;
    }
    ntest::generate_report("validation");
  }
  catch (std::runtime_error const &err)
  {
    std::cerr << "fatal: " << err.what() << '\n';
    return 1;
  }
  catch (...)
  {
    std::cerr << "fatal: unknown error\n";
    return 2;
  }

  return 0;
}
