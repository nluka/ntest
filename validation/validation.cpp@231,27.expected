(Alice Smith, 22)
(Joe Schmoe, 100)
(Rob Robertson, 53)
